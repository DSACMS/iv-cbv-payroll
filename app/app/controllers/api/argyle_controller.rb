class Api::ArgyleController < ApplicationController
  before_action :set_cbv_flow, :resume_previous_argyle_account_connection
  after_action :track_event

  # This API endpoint is used to fetch a `user_token` to allow the user to open
  # the Argyle modal.
  #
  # @see https://docs.argyle.com/link/user-tokens
  def create
    is_sandbox_environment = agency_config[@cbv_flow.client_agency_id].argyle_environment == "sandbox"
    user_token = if @cbv_flow.argyle_user_id.blank?
                   response = argyle.create_user(@cbv_flow.end_user_id)

                   # Store the argyle_user_id to allow us to associate incoming webhooks with
                   # this CbvFlow.
                   @cbv_flow.update(argyle_user_id: response["id"])

                   response["user_token"]
                 else
                   # If the user has already been created in Argyle, let's just
                   # make them a new link token with the same user.
                   response = argyle.create_user_token(@cbv_flow.argyle_user_id)
                   response["user_token"]
                 end

    render json: {
      status: :ok,
      isSandbox: is_sandbox_environment,
      flowId: Aggregators::Sdk::ArgyleService::FLOW_ID,
      user: {
        user_token: user_token
      }
    }
  end

  private

  def set_cbv_flow
    @cbv_flow = CbvFlow.find_by(id: session[:cbv_flow_id])
    redirect_to(root_url(cbv_flow_timeout: true)) unless @cbv_flow
  end

  # Redirect if the user is attempting connect a previously connected account
  def resume_previous_argyle_account_connection
    item_id = params[:item_id]
    unless item_id.present?
      return render json: { status: :error, message: "Invalid item_id" }, status: :unprocessable_entity
    end

    return if item_id == "item_000026933" # If it's an ADP item, allow user to create a new connection.

    # If the user ID is not yet set, there is no previous argyle session to resume
    return unless @cbv_flow.argyle_user_id.present?

    # Find any previous Argyle connections to this item
    connected_argyle_accounts = argyle.fetch_accounts_api(user: @cbv_flow.argyle_user_id, item: item_id)["results"]
    return unless connected_argyle_accounts.any?

    # Find the PayrollAccount object (if we have received an accounts.connected webhook)
    argyle_account = connected_argyle_accounts.first
    payroll_account = @cbv_flow.payroll_accounts.find_by(pinwheel_account_id: argyle_account["id"])
    return unless payroll_account.present?

    # If we've made it here, there is a previous connection to that item.
    #
    # Redirect to the proper place based on the status of the connection.
    if payroll_account.sync_succeeded? || payroll_account.sync_failed?
      redirect_to cbv_flow_payment_details_path(user: { account_id: payroll_account.pinwheel_account_id })
    elsif payroll_account.job_succeeded?("accounts")
      # Sync is in progress, and the user has successfully connected their account.
      redirect_to cbv_flow_synchronizations_path(user: { account_id: payroll_account.pinwheel_account_id })
    end
  end

  def track_event
    return unless @cbv_flow.present?

    event_logger.track(TrackEvent::ApplicantBeganLinkingEmployer, request, {
      time: Time.now.to_i,
      cbv_flow_id: @cbv_flow.id,
      cbv_applicant_id: @cbv_flow.cbv_applicant_id,
      client_agency_id: @cbv_flow.client_agency_id,
      invitation_id: @cbv_flow.cbv_flow_invitation_id,
      item_id: params[:item_id],
      aggregator_name: "argyle"
    })
  end

  def argyle
    argyle_for(@cbv_flow)
  end
end
