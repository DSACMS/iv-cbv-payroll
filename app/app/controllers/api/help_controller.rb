class Api::HelpController < ApplicationController
  EVENT_NAMES = %w[
    ApplicantOpenedHelpModal
  ]

  def user_action
    base_attributes = {
      timestamp: Time.now.to_i
    }

    event_name = user_action_params[:event_name]
    event_attributes = base_attributes.merge(source: user_action_params[:source])

    if EVENT_NAMES.include?(event_name)
      event_logger.track(
        event_name,
        request,
        event_attributes.to_h
      )
    else
      raise "Unknown Event Type #{event_name.inspect}"
    end

    render json: { status: :ok }
  rescue => ex
    raise ex unless Rails.env.production?

    Rails.logger.error "Unable to process user action: #{ex}"
    render json: { status: :error }, status: :unprocessable_entity
  end

  private

  def user_action_params
    params.permit(:event_name, :source)
  end
end 