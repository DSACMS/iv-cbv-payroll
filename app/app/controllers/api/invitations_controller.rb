class Api::InvitationsController < ApplicationController
  skip_forgery_protection
  wrap_parameters false

  before_action :authenticate

  def create
    @cbv_flow_invitation = CbvInvitationService.new(event_logger)
      .invite(cbv_flow_invitation_params, @current_user, delivery_method: nil)

    errors = @cbv_flow_invitation.errors
    if errors.any?
      return render json: errors_to_json(errors), status: :unprocessable_entity
    end

    render json: {
      tokenized_url: @cbv_flow_invitation.to_url,
      expiration_date: @cbv_flow_invitation.expires_at_local,
      language: @cbv_flow_invitation.language
      # TODO: Determine if we actually want to echo this data back, maybe should be config param?
      # agency_partner_metadata: allowed_metadata_params
    }, status: :created
  end

  private

  def cbv_flow_invitation_params
    client_agency_id = @current_user.client_agency_id

    # Permit top-level params of the invitation itself, while merging back in
    # the allowed applicant attributes.
    permitted = params.without(:client_agency_id, :agency_partner_metadata).permit(:language, :email_address, :user_id)
    permitted.deep_merge!(
      client_agency_id: client_agency_id,
      email_address: @current_user.email,
      cbv_applicant_attributes: {
        client_agency_id: client_agency_id,
        **allowed_metadata_params.permit!
      }
    )
  end

  def allowed_metadata_params
    # Allow params in the VALID_ATTRIBUTES array for the relevant agency
    # CbvApplicant subclass.
    ActionController::Parameters.new(
      CbvApplicant.build_agency_partner_metadata(@current_user.client_agency_id) { |attr| params[:agency_partner_metadata][attr] }
    )
  end

  def authenticate
    authenticate_or_request_with_http_token do |token, options|
      @current_user = User.find_by_access_token(token)
    end
  end

  def errors_to_json(errors)
    # Generates a Hash of attribute => error_message and translates the
    # internal names of objects (cbv_applicant) to the external names
    # (agency_partner_metadata)
    error_messages = errors.map do |error|
      next if error.attribute == :cbv_applicant

      error_message = error.message

      case error
      when ActiveModel::NestedError
        prefix, attribute_name = error.attribute.to_s.split(".")
        prefix = "agency_partner_metadata" if prefix == "cbv_applicant"

        { field: "#{prefix}.#{attribute_name}", message: error_message }
      else
        { field: error.attribute, message: error_message }
      end
    end.compact

    { errors: error_messages }
  end
end
