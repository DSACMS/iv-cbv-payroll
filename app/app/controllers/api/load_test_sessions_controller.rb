class Api::LoadTestSessionsController < ApplicationController
  include NonProductionAccessible

  skip_forgery_protection

  # Only allow in non-production environments (development/test/demo)
  before_action :ensure_non_production_environment

  def create
    client_agency_id = params[:client_agency_id] || "sandbox"
    scenario = params[:scenario] || "synced"

    # Validate client_agency_id
    unless Rails.application.config.client_agencies.client_agency_ids.include?(client_agency_id)
      return render json: { error: "Invalid client_agency_id" }, status: :unprocessable_entity
    end

    # Create test data based on scenario
    cbv_flow, account_id = case scenario
                           when "synced"
                             create_synced_flow(client_agency_id)
                           when "pending"
                             create_pending_flow(client_agency_id)
                           when "failed"
                             create_failed_flow(client_agency_id)
                           else
                             return render json: { error: "Invalid scenario: #{scenario}" }, status: :unprocessable_entity
                           end

    # Set session using Rails' session mechanism (Rails will encrypt the cookie)
    session[:cbv_flow_id] = cbv_flow.id

    render json: {
      success: true,
      cbv_flow_id: cbv_flow.id,
      account_id: account_id,
      client_agency_id: client_agency_id,
      scenario: scenario,
      csrf_token: form_authenticity_token,
      message: "Session created. Cookie will be set in Set-Cookie header."
    }, status: :created
  end

  private

  def argyle_account_id
    # hard coded to bob's id to match the mock api service
    "019571bc-2f60-3955-d972-dbadfe0913a8"
  end

  def ensure_non_production_environment
    unless is_not_production?
      render json: { error: "This endpoint is only available in non-production environments" }, status: :forbidden
    end
  end

  def create_synced_flow(client_agency_id)
    cbv_applicant = CbvApplicant.create!(client_agency_id: client_agency_id)
    cbv_flow = CbvFlow.create!(
      client_agency_id: client_agency_id,
      cbv_applicant: cbv_applicant,
      consented_to_authorized_use_at: Time.current
    )

    # Create fully synced payroll account
    payroll_account = PayrollAccount::Argyle.create!(
      cbv_flow: cbv_flow,
      aggregator_account_id: argyle_account_id,
      pinwheel_account_id: argyle_account_id,
      supported_jobs: %w[accounts income paystubs employment identity],
      synchronization_status: :succeeded
    )

    # Create successful webhook events matching Argyle's actual event names
    # See: Aggregators::Webhooks::Argyle::SUBSCRIBED_WEBHOOK_EVENTS
    [
      { event_name: "accounts.connected", event_outcome: "success" },       # accounts job
      { event_name: "identities.added", event_outcome: "success" },        # identity + income jobs
      { event_name: "paystubs.fully_synced", event_outcome: "success" }    # paystubs + employment jobs
    ].each do |event|
      WebhookEvent.create!(
        payroll_account: payroll_account,
        event_name: event[:event_name],
        event_outcome: event[:event_outcome]
      )
    end

    [ cbv_flow, argyle_account_id ]
  end

  def create_pending_flow(client_agency_id)
    cbv_applicant = CbvApplicant.create!(client_agency_id: client_agency_id)
    cbv_flow = CbvFlow.create!(
      client_agency_id: client_agency_id,
      cbv_applicant: cbv_applicant,
      consented_to_authorized_use_at: Time.current
    )

    # Create pending payroll account
    payroll_account = PayrollAccount::Argyle.create!(
      cbv_flow: cbv_flow,
      pinwheel_account_id: argyle_account_id,
      aggregator_account_id: argyle_account_id,
      supported_jobs: %w[accounts income paystubs employment identity],
      synchronization_status: :in_progress
    )

    # Create initial webhook event (account connected, but sync still in progress)
    WebhookEvent.create!(
      payroll_account: payroll_account,
      event_name: "accounts.connected",
      event_outcome: "success"
    )

    [ cbv_flow, argyle_account_id ]
  end

  def create_failed_flow(client_agency_id)
    cbv_applicant = CbvApplicant.create!(client_agency_id: client_agency_id)
    cbv_flow = CbvFlow.create!(
      client_agency_id: client_agency_id,
      cbv_applicant: cbv_applicant,
      consented_to_authorized_use_at: Time.current
    )

    # Create failed payroll account
    payroll_account = PayrollAccount::Argyle.create!(
      cbv_flow: cbv_flow,
      pinwheel_account_id: argyle_account_id,
      aggregator_account_id: argyle_account_id,
      supported_jobs: %w[accounts income paystubs employment identity],
      synchronization_status: :failed
    )

    # Create failed webhook events - paystubs job failed
    [
      { event_name: "accounts.connected", event_outcome: "success" },
      { event_name: "paystubs.fully_synced", event_outcome: "error" }  # This marks paystubs/employment as failed
    ].each do |event|
      WebhookEvent.create!(
        payroll_account: payroll_account,
        event_name: event[:event_name],
        event_outcome: event[:event_outcome]
      )
    end

    [ cbv_flow, argyle_account_id ]
  end
end
