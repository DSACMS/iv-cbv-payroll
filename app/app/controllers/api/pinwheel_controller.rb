class Api::PinwheelController < ApplicationController
  after_action :track_event, only: :create_token

  # run the token here with the included employer/payroll provider id
  def create_token
    @cbv_flow = CbvFlow.find(session[:cbv_flow_id])
    pinwheel = pinwheel_for(@cbv_flow)
    token_response = pinwheel.create_link_token(
      response_type: token_params[:response_type],
      id: token_params[:id],
      end_user_id: @cbv_flow.end_user_id,
      language: token_params[:locale]
    )
    token = token_response["data"]["token"]

    render json: { status: :ok, token: token }
  end

  private

  def token_params
    params.require(:pinwheel).permit(:response_type, :id, :locale)
  end

  def track_event
    event_logger.track("ApplicantBeganLinkingEmployer", request, {
      time: Time.now.to_i,
      cbv_flow_id: @cbv_flow.id,
      cbv_applicant_id: @cbv_flow.cbv_applicant_id,
      client_agency_id: @cbv_flow.client_agency_id,
      invitation_id: @cbv_flow.cbv_flow_invitation_id,
      response_type: token_params[:response_type],
      item_id: token_params[:id],
      aggregator_name: "pinwheel"
    })
  rescue => ex
    Rails.logger.error "Unable to track event (ApplicantBeganLinkingEmployer): #{ex}"
  end
end
