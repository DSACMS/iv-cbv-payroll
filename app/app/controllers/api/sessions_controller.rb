class Api::SessionsController < ApplicationController
  # Skip CSRF protection for API calls if you're using protect_from_forgery
  # skip_before_action :verify_authenticity_token, only: [:extend]
  
  def extend
    Rails.logger.info "API::SessionsController#extend called"
    
    # Reset the Devise timeout timer
    request.env['devise.skip_timeout'] = true
    
    if current_user
      Rails.logger.info "Current user found: #{current_user.id}"
      current_user.remember_me! if current_user.respond_to?(:remember_me!)
      current_user.remember_me = true if current_user.respond_to?(:remember_me=)
      
      # Update the last_request_at timestamp to extend the Devise timeout
      if current_user.respond_to?(:update_tracked_fields!)
        current_user.update_tracked_fields!(request)
      end
      
      # Force sign in to reset the Devise timeout
      sign_in(current_user, force: true)
      
      respond_to do |format|
        format.json { 
          Rails.logger.info "Responding with JSON"
          render json: { success: true, message: "Session extended successfully" } 
        }
        format.html { 
          Rails.logger.info "Responding with HTML redirect"
          redirect_back(fallback_location: root_path) 
        }
        format.any { 
          Rails.logger.info "Responding with 406 Not Acceptable"
          head :not_acceptable 
        }
      end
    elsif session[:cbv_flow_id].present?
      # Handle CBV flow session extension
      Rails.logger.info "Extending session for CBV flow: #{session[:cbv_flow_id]}"
      
      # Reset the session timestamp
      request.env["rack.session.options"][:expire_after] = 30.minutes
      
      respond_to do |format|
        format.json { 
          Rails.logger.info "Responding with JSON for CBV flow"
          render json: { success: true, message: "Session extended successfully" } 
        }
        format.html { 
          Rails.logger.info "Responding with HTML redirect for CBV flow"
          redirect_back(fallback_location: root_path) 
        }
        format.any { 
          Rails.logger.info "Responding with 406 Not Acceptable for CBV flow"
          head :not_acceptable 
        }
      end
    else
      Rails.logger.warn "No current user or CBV flow found"
      respond_to do |format|
        format.json { render json: { success: false, error: "No active session" }, status: :unauthorized }
        format.html { redirect_to new_user_session_path }
        format.any { head :unauthorized }
      end
    end
  rescue => e
    Rails.logger.error "Error in extend session: #{e.message}"
    Rails.logger.error e.backtrace.join("\n")
    respond_to do |format|
      format.json { render json: { success: false, error: e.message }, status: :internal_server_error }
      format.html { redirect_to root_path, alert: "An error occurred while extending your session." }
      format.any { head :internal_server_error }
    end
  end
end 