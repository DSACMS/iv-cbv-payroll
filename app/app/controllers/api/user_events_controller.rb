class Api::UserEventsController < ApplicationController
  EVENT_NAMES = %w[
    ApplicantOpenedHelpModal
    ApplicantCopiedInvitationLink
    ApplicantSelectedEmployerOrPlatformItem
    ApplicantAttemptedClosingPinwheelModal
    ApplicantAttemptedPinwheelLogin
    ApplicantClosedPinwheelModal
    ApplicantEncounteredPinwheelError
    ApplicantViewedPinwheelDefaultProviderSearch
    ApplicantViewedPinwheelLoginPage
    ApplicantViewedPinwheelProviderConfirmation
    ApplicantSucceededWithPinwheelLogin
    ApplicantSucceededWithArgyleLogin
    ApplicantCreatedArgyleAccount
    ApplicantEncounteredArgyleAccountError
    ApplicantRemovedArgyleAccount
    ApplicantClosedArgyleModal
    ApplicantEncounteredArgyleError
    ApplicantEncounteredArgyleTokenExpired
    ApplicantEncounteredModalAdapterError
    ApplicantViewedArgyleProviderConfirmation
    ApplicantViewedArgyleLoginPage
    ApplicantAttemptedArgyleLogin
    ApplicantViewedArgyleDefaultProviderSearch
    ApplicantAccessedArgyleModalMFAScreen
    ApplicantEncounteredArgyleInvalidCredentialsLoginError
    ApplicantEncounteredArgyleAuthRequiredLoginError
    ApplicantEncounteredArgyleConnectionUnavailableLoginError
    ApplicantEncounteredArgyleExpiredCredentialsLoginError
    ApplicantEncounteredArgyleInvalidAuthLoginError
    ApplicantEncounteredArgyleMfaCanceledLoginError
    ApplicantUpdatedArgyleSearchTerm
    ApplicantManuallySwitchedLanguage
    ApplicantConsentedToTerms
    ApplicantViewedHelpText
  ]

  def user_action
    base_attributes = {
      time: Time.now.to_i
    }

    if session[:cbv_flow_id].present?
      @cbv_flow = CbvFlow.find(session[:cbv_flow_id])

      base_attributes.merge!({
        cbv_flow_id: @cbv_flow.id,
        cbv_applicant_id: @cbv_flow.cbv_applicant_id,
        client_agency_id: @cbv_flow.client_agency_id,
        invitation_id: @cbv_flow.cbv_flow_invitation_id
      })
    end

    event_attributes = (user_action_params[:attributes] || {}).merge(base_attributes)
    event_name = user_action_params[:event_name]

    if EVENT_NAMES.include?(event_name)
      event_logger.track(
        event_name,
        request,
        event_attributes.to_h
      )
    else
      raise "Unknown Event Type #{event_name.inspect}"
    end

    render json: { status: :ok }

  rescue => ex
    raise ex unless Rails.env.production?

    Rails.logger.error "Unable to process user action: #{ex}"
    render json: { status: :error }, status: :unprocessable_entity
  end

  private

  def user_action_params
    params.fetch(:events, {}).permit(:event_name, attributes: {})
  end
end
