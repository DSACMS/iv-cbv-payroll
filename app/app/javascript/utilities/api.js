import CSRF from './csrf';

export const PINWHEEL_USER_ACTION = '/api/pinwheel/user_action';
export const PINWHEEL_TOKENS_GENERATE = '/api/pinwheel/tokens';

export const trackUserAction = (eventName, attributes, scope="pinwheel") => {
  return fetch(PINWHEEL_USER_ACTION, {
    method: 'post',
    headers: {
      'X-CSRF-Token': CSRF.token,
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({ [scope]: { event_name: eventName, attributes } }),
  }).then(response => response.json());
};

export const fetchToken = (response_type, id, locale) => {
  return fetch(PINWHEEL_TOKENS_GENERATE, {
    method: 'post',
    headers: {
      'X-CSRF-Token': CSRF.token,
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({ response_type, id, locale }),
  }).then(response => response.json());
};