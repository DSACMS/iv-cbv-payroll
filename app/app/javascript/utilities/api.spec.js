import { vi, describe, beforeEach, afterEach, it, expect } from 'vitest'
import * as api from './api'

global.fetch = vi.fn()

function createFetchResponse(data) {
  return { json: () => new Promise((resolve) => resolve(data)) }
}
 // Here we tell Vitest to mock fetch on the `window` object.
describe('trackUserAction', () => {
    beforeEach(async() => {
        // Mock the fetch function.
        const mockResponse = {
            "pinwheel": {
                "event_name": "Event",
                "attributes": {}
            }
        };
        fetch.mockResolvedValue(createFetchResponse(mockResponse))
    })

    it('sends a post request to the user_action endpoint', async () => {
        const data = await api.trackUserAction("Event", {})

        // Check that fetch was called exactly once
        expect(fetch).toHaveBeenCalledTimes(1);
        expect(fetch.mock.calls[0][0]).toBe('/api/pinwheel/user_action')
        expect(fetch.mock.calls[0][1]['method']).toBe('post')
    })

    it('includes CSRV and Content-Type headers', async () => {
        const data = await api.trackUserAction("Event", {})
        expect(fetch).toHaveBeenCalledTimes(1);
        expect(fetch.mock.calls[0][1]).toHaveProperty('headers')
        expect(fetch.mock.calls[0][1]['headers']).toHaveProperty('X-CSRF-Token')
        expect(fetch.mock.calls[0][1]['headers']).toHaveProperty('Content-Type')
        expect(fetch.mock.calls[0][1]['headers']['Content-Type']).toEqual('application/json')
    })

    it('has expected request body', async() => {
        const data = await api.trackUserAction("Event", {})
        expect(fetch.mock.calls[0][1]['body']).toMatchSnapshot()
    })
    it('has expected response payload', async() => {
        const data = await api.trackUserAction("Event", {})
        expect(data).toMatchSnapshot()
    })
        

    afterEach(() => {
        fetch.mockReset()
    })
})
