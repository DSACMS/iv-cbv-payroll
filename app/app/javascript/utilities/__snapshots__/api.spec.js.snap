// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`trackUserAction > has expected request body 1`] = `"{"pinwheel":{"event_name":"Event","attributes":{}}}"`;

exports[`trackUserAction > has expected response payload 1`] = `
{
  "pinwheel": {
    "attributes": {},
    "event_name": "Event",
  },
}
`;
